test/unit/hpa_central.c

```diff
@@ -0,0 +1,452 @@
+	assert_ptr_not_null(edata, "");
+	edata_init(edata, base_ind_get(base), (void *)addr,
+	    size, /* slab */ false, /* szind_t */ SC_NSIZES, /* sn */ 0,
+	    extent_state_active, /* zeroed */ true, /* comitted */ true,
+	    EXTENT_PAI_HPA, /* is_head */ true);
+	return edata;
+}
+
+static void
+edata_expect_alloc(base_t *base, edata_t *edata, uintptr_t addr, size_t size) {
+	expect_ptr_not_null(edata, "Alloc should have succeeded");
+	expect_u_eq(base_ind_get(base), edata_arena_ind_get(edata), "");
+	expect_u_eq(SC_NSIZES, edata_szind_get_maybe_invalid(edata), "");
+	expect_d_eq(extent_state_active, edata_state_get(edata), "");
+	assert_ptr_eq((void *)addr, edata_base_get(edata), "");
+	assert_zu_eq(size, edata_size_get(edata), "");
+}
+
+
+TEST_BEGIN(test_empty) {
+	hpa_central_t *central;
+	base_t *base;
+	create_test_data(&central, &base);
+
+	edata_t *edata;
+
+	edata = hpa_central_alloc_reuse(TSDN_NULL, central, PAGE, PAGE);
+	expect_ptr_null(edata, "Empty allocator succeed in its allocation");
+
+	edata = hpa_central_alloc_reuse(TSDN_NULL, central, PAGE, 2 * PAGE);
+	expect_ptr_null(edata, "Empty allocator succeed in its allocation");
+
+	edata = hpa_central_alloc_reuse(TSDN_NULL, central, PAGE, 8 * PAGE);
+	expect_ptr_null(edata, "Empty allocator succeed in its allocation");
+
+	edata = hpa_central_alloc_reuse(TSDN_NULL, central, 4 * PAGE, 8 * PAGE);
+	expect_ptr_null(edata, "Empty allocator succeed in its allocation");
+
+	destroy_test_data(central);
+}
+TEST_END
+
+TEST_BEGIN(test_first_fit_simple) {
+	hpa_central_t *central;
+	base_t *base;
+	create_test_data(&central, &base);
+
+	edata_t *edata1 = test_edata(base, 10 * PAGE, 10 * PAGE);
+	bool err = hpa_central_alloc_grow(TSDN_NULL, central, PAGE, edata1);
+	expect_false(err, "Unexpected grow failure");
+	edata_expect_alloc(base, edata1, 10 * PAGE, PAGE);
+
+	edata_t *edata2 = test_edata(base, 4 * PAGE, 1 * PAGE);
+	err = hpa_central_alloc_grow(TSDN_NULL, central, PAGE, edata2);
+	expect_false(err, "Unexpected grow failure");
+	edata_expect_alloc(base, edata2, 4 * PAGE, PAGE);
+
+	hpa_central_dalloc(TSDN_NULL, central, edata2);
+
+	/*
+	 * Even though there's a lower-addressed extent that a by-size search
+	 * will find earlier, we should still pick the earlier one.
+	 */
+	edata_t *edata3 = hpa_central_alloc_reuse(TSDN_NULL, central, PAGE, PAGE);
+	/*
+	 * Recall there's still an active page at the beginning of the extent
+	 * added at 10 * PAGE; the next allocation from it should be at 11 *
+	 * PAGE.
+	 */
+	edata_expect_alloc(base, edata3, 11 * PAGE, PAGE);
+
+	destroy_test_data(central);
+}
+TEST_END
+
+TEST_BEGIN(test_first_fit_large_goal) {
+	/*
+	 * Fragmentation avoidance is more important in the HPA than giving the
+	 * user their preferred amount of space, since we expect the average
+	 * unused extent to be more costly (PAC extents can get purged away
+	 * easily at any granularity; HPA extents are much more difficult to
+	 * purge away if they get stranded).
```

Comment:
The explanation here is pretty helpful; maybe move it to the source instead?

Reply:
Will do.